Prevent RTL text from mirrorring around styled characters for on-screen keyboard

AhmedAdelAbouzeid opened the pull request:
## Description of changes and context
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->
This PR is a follow-up for for #17692 and includes 3 main changes:

1- Revert the character used for cursor blinking for on-screen keyboard to the original as replacing it with a basic space causes text to move sideways during blinking for the default font.

2- Instead of replacing the styled character, BidiTransform() function was re-written such that calls to FriBidi are performed for a complete line at a time rather than breaking it for differently styled sections (RTL text mirroring around styled characters occurred due to this line splitting).
Now, the styling data for the whole line is saved and re-applied after FriBidi call. To support this functionality a mapping is created between logical and visual strings as text that was originally styled may not be in the same location after FriBidi call. The function BidiFlip() was also modified for this purpose.

3- Logical to visual string mapping was added as an optional parameter for functions wrapping FriBidi calls to support functionality needed in point 2.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->
The changes were tested using Kodi on Windows 10 x64 (1909)
The on-screen keyboard was tested using a RTL language (Arabic) and a LTR language (English) and behave correctly for both languages after the changes.

## Screenshots (if appropriate):
![Kodi_KB_AR_CR](https://user-images.githubusercontent.com/63805744/80705744-3df89080-8ae7-11ea-944c-831e59a0fda2.png)


## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

phunkyfish commented:
@AlwinEsch this PR should solve the issue with the text moving. Do you want to check if it’s ok for you?

AlwinEsch reviewed (approved):
Have checked with different languages and fonts, was ok :+1:

phunkyfish commented on xbmc/guilib/GUITextLayout.cpp:
```diff
@@ -261,51 +261,52 @@ void CGUITextLayout::BidiTransform(std::vector<CGUIString> &lines, bool forceLTR
   for (unsigned int i=0; i<lines.size(); i++)
   {
     CGUIString &line = lines[i];
+	unsigned int lineLength = line.m_text.size();
```
indentation wrong

phunkyfish commented on xbmc/guilib/GUITextLayout.cpp:
```diff
@@ -261,51 +261,52 @@ void CGUITextLayout::BidiTransform(std::vector<CGUIString> &lines, bool forceLTR
   {
     CGUIString &line = lines[i];
+	unsigned int lineLength = line.m_text.size();
+	std::wstring logicalText;
```
indentation wrong

phunkyfish commented on xbmc/guilib/GUITextLayout.cpp:
```diff
@@ -261,51 +261,52 @@ void CGUITextLayout::BidiTransform(std::vector<CGUIString> &lines, bool forceLTR
     }
+	
+	// Allocate memory for logical to visual map and call bidi
+    int* logicalToVisualMap = (int*)malloc((lineLength + 1) * sizeof(int));
```
Why malloc? is there a reason not to use `new`?

`int* logicalToVisualMap = new int[lineLength + 1];`

phunkyfish commented on xbmc/guilib/GUITextLayout.cpp:
```diff
@@ -261,51 +261,52 @@ void CGUITextLayout::BidiTransform(std::vector<CGUIString> &lines, bool forceLTR
+    styledVisualText.reserve(lineLength);
+
+    // If malloc failed, fallback to text with no styling
+    if (logicalToVisualMap == NULL)
```
`nullptr`

phunkyfish commented on xbmc/guilib/GUITextLayout.cpp:
```diff
@@ -261,51 +261,52 @@ void CGUITextLayout::BidiTransform(std::vector<CGUIString> &lines, bool forceLTR
+
+    // If malloc failed, fallback to text with no styling
+    if (logicalToVisualMap == NULL)
+      for (unsigned int j = 0; j < visualText.size(); j++)
+        styledVisualText.push_back(visualText[j]);
```
Surrounded by brackets

phunkyfish commented on xbmc/guilib/GUITextLayout.cpp:
```diff
@@ -261,51 +261,52 @@ void CGUITextLayout::BidiTransform(std::vector<CGUIString> &lines, bool forceLTR
+      for (unsigned int j = 0; j < visualText.size(); j++)
+        styledVisualText.push_back(visualText[j]);
+    else
+      for (unsigned int j = 0; j < visualText.size(); j++)
+        styledVisualText.push_back(style[logicalToVisualMap[j]] | visualText[j]);
```
Surrounded by brackets

phunkyfish commented on xbmc/guilib/GUITextLayout.cpp:
```diff
@@ -261,51 +261,52 @@ void CGUITextLayout::BidiTransform(std::vector<CGUIString> &lines, bool forceLTR
+  // Convert to utf32, call bidi then convert the result back to utf16
+  g_charsetConverter.wToUtf32(text, utf32logical);
+  g_charsetConverter.utf32logicalToVisualBiDi(utf32logical, utf32visual, false, false, logicalToVisualMap);
+  g_charsetConverter.utf32ToW(utf32visual,visualText);
```
space after comma

phunkyfish commented on xbmc/guilib/GUITextLayout.cpp:
```diff
@@ -261,51 +261,52 @@ void CGUITextLayout::BidiTransform(std::vector<CGUIString> &lines, bool forceLTR
-      for (unsigned int j = 0; j < sectionFlipped.size(); j++)
-        flippedText.push_back(sectionStyle | sectionFlipped[j]);
-    }
+    free(logicalToVisualMap);
```
`delete[] logicalToVisualMap;`

phunkyfish commented on xbmc/guilib/GUITextLayout.cpp:
```diff
@@ -261,51 +261,52 @@ void CGUITextLayout::BidiTransform(std::vector<CGUIString> &lines, bool forceLTR
 }
 
-std::wstring CGUITextLayout::BidiFlip(const std::wstring &text, bool forceLTRReadingOrder)
+std::wstring CGUITextLayout::BidiFlip(const std::wstring &text, bool forceLTRReadingOrder, int* logicalToVisualMap /*=NULL*/)
```
space after `=`

phunkyfish commented on xbmc/guilib/GUITextLayout.h:
```diff
@@ -108,7 +108,7 @@ class CGUITextLayout
   void WrapText(const vecText &text, float maxWidth);
   static void BidiTransform(std::vector<CGUIString> &lines, bool forceLTRReadingOrder);
-  static std::wstring BidiFlip(const std::wstring &text, bool forceLTRReadingOrder);
+  static std::wstring BidiFlip(const std::wstring &text, bool forceLTRReadingOrder, int* logicalToVisualMap = NULL);
```
`nullptr`

phunkyfish commented on xbmc/utils/CharsetConverter.cpp:
```diff
@@ -266,7 +266,7 @@ enum StdConversionType /* Keep it in sync with CCharsetConverter::CInnerConverte
 {
 public:
-  static bool logicalToVisualBiDi(const std::u32string& stringSrc, std::u32string& stringDst, FriBidiCharType base = FRIBIDI_TYPE_LTR, const bool failOnBadString = false);
+  static bool logicalToVisualBiDi(const std::u32string& stringSrc, std::u32string& stringDst, FriBidiCharType base = FRIBIDI_TYPE_LTR, const bool failOnBadString = false, int* logicalToVisualMap = NULL);
```
`nullptr` and line length needs to be less than 100 chars

phunkyfish commented on xbmc/utils/CharsetConverter.cpp:
```diff
@@ -676,9 +676,9 @@ bool CCharsetConverter::utf32ToW(const std::u32string& utf32StringSrc, std::wstr
 }
 
-bool CCharsetConverter::utf32logicalToVisualBiDi(const std::u32string& logicalStringSrc, std::u32string& visualStringDst, bool forceLTRReadingOrder /*= false*/, bool failOnBadString /*= false*/)
+bool CCharsetConverter::utf32logicalToVisualBiDi(const std::u32string& logicalStringSrc, std::u32string& visualStringDst, bool forceLTRReadingOrder /*= false*/, bool failOnBadString /*= false*/, int* logicalToVisualMap /*= NULL*/)
```
Line too long

phunkyfish commented on xbmc/utils/CharsetConverter.cpp:
```diff
@@ -676,9 +676,9 @@ bool CCharsetConverter::utf32ToW(const std::u32string& utf32StringSrc, std::wstr
+bool CCharsetConverter::utf32logicalToVisualBiDi(const std::u32string& logicalStringSrc, std::u32string& visualStringDst, bool forceLTRReadingOrder /*= false*/, bool failOnBadString /*= false*/, int* logicalToVisualMap /*= NULL*/)
 {
-  return CInnerConverter::logicalToVisualBiDi(logicalStringSrc, visualStringDst, forceLTRReadingOrder ? FRIBIDI_TYPE_LTR : FRIBIDI_TYPE_PDF, failOnBadString);
+  return CInnerConverter::logicalToVisualBiDi(logicalStringSrc, visualStringDst, forceLTRReadingOrder ? FRIBIDI_TYPE_LTR : FRIBIDI_TYPE_PDF, failOnBadString, logicalToVisualMap);
```
line too long

phunkyfish commented on xbmc/utils/CharsetConverter.h:
```diff
@@ -98,9 +98,10 @@ class CCharsetConverter : public ISettingCallback
    * @return true on success, false otherwise
    */
-  static bool utf32logicalToVisualBiDi(const std::u32string& logicalStringSrc, std::u32string& visualStringDst, bool forceLTRReadingOrder = false, bool failOnBadString = false);
+  static bool utf32logicalToVisualBiDi(const std::u32string& logicalStringSrc, std::u32string& visualStringDst, bool forceLTRReadingOrder = false, bool failOnBadString = false, int* logicalToVisualMap = NULL);
```
line too long, `nullptr`

AhmedAdelAbouzeid commented:
Thanks a lot for reviewing.
I noted and applied the requested code changes. Please re-check, Thanks!

phunkyfish commented on xbmc/guilib/GUITextLayout.cpp:
```diff
@@ -261,51 +261,61 @@ void CGUITextLayout::BidiTransform(std::vector<CGUIString> &lines, bool forceLTR
+    styledVisualText.reserve(lineLength);
+
+    // If memory allocation failed, fallback to text with no styling
+    if (logicalToVisualMap == nullptr)
```
`if (! logicalToVisualMap)`

phunkyfish commented on xbmc/guilib/GUITextLayout.cpp:
```diff
@@ -261,51 +261,61 @@ void CGUITextLayout::BidiTransform(std::vector<CGUIString> &lines, bool forceLTR
 
-std::wstring CGUITextLayout::BidiFlip(const std::wstring &text, bool forceLTRReadingOrder)
+std::wstring CGUITextLayout::BidiFlip(const std::wstring &text, bool forceLTRReadingOrder,
+  int* logicalToVisualMap /*= nullptr*/)
```
align with other arguements, e.g.

```
std::wstring CGUITextLayout::BidiFlip(const std::wstring &text, bool forceLTRReadingOrder,
                                      int* logicalToVisualMap /*= nullptr*/)
```

phunkyfish commented on xbmc/guilib/GUITextLayout.h:
```diff
@@ -108,7 +108,8 @@ class CGUITextLayout
   static void BidiTransform(std::vector<CGUIString> &lines, bool forceLTRReadingOrder);
-  static std::wstring BidiFlip(const std::wstring &text, bool forceLTRReadingOrder);
+  static std::wstring BidiFlip(const std::wstring &text, bool forceLTRReadingOrder,
+    int* logicalToVisualMap = nullptr);
```
Align to other arguments like previous example.

phunkyfish commented on xbmc/utils/CharsetConverter.cpp:
```diff
@@ -266,7 +266,9 @@ enum StdConversionType /* Keep it in sync with CCharsetConverter::CInnerConverte
 public:
-  static bool logicalToVisualBiDi(const std::u32string& stringSrc, std::u32string& stringDst, FriBidiCharType base = FRIBIDI_TYPE_LTR, const bool failOnBadString = false);
+  static bool logicalToVisualBiDi(const std::u32string& stringSrc, std::u32string& stringDst,
+    FriBidiCharType base = FRIBIDI_TYPE_LTR, const bool failOnBadString = false,
+    int* logicalToVisualMap = nullptr);
```
Maybe like this, we have enough width to align to the function brackets:

```cpp
static bool logicalToVisualBiDi(const std::u32string& stringSrc,
                                std::u32string& stringDst,
                                FriBidiCharType base = FRIBIDI_TYPE_LTR,
                                const bool failOnBadString = false,
                                int* logicalToVisualMap = nullptr);
```

phunkyfish commented on xbmc/utils/CharsetConverter.cpp:
```diff
@@ -462,7 +464,9 @@ bool CCharsetConverter::CInnerConverter::convert(iconv_t type, int multiplier, c
 
-bool CCharsetConverter::CInnerConverter::logicalToVisualBiDi(const std::u32string& stringSrc, std::u32string& stringDst, FriBidiCharType base /*= FRIBIDI_TYPE_LTR*/, const bool failOnBadString /*= false*/)
+bool CCharsetConverter::CInnerConverter::logicalToVisualBiDi(const std::u32string& stringSrc,
+  std::u32string& stringDst, FriBidiCharType base /*= FRIBIDI_TYPE_LTR*/,
+  const bool failOnBadString /*= false*/, int* logicalToVisualMap /*= nullptr*/)
```
Like this. We need to use a 4 space continuance as aligning to the function brackets breaks the 100 char width limit.

```cpp
bool CCharsetConverter::CInnerConverter::logicalToVisualBiDi(
    const std::u32string& stringSrc,
    std::u32string& stringDst,
    FriBidiCharType base /*= FRIBIDI_TYPE_LTR*/,
    const bool failOnBadString /*= false*/,
    int* logicalToVisualMap /*= nullptr*/)
```

phunkyfish commented on xbmc/utils/CharsetConverter.cpp:
```diff
@@ -495,7 +499,8 @@ bool CCharsetConverter::CInnerConverter::logicalToVisualBiDi(const std::u32strin
     bool bidiFailed = false;
     FriBidiCharType baseCopy = base; // preserve same value for all lines, required because fribidi_log2vis will modify parameter value
-    if (fribidi_log2vis((const FriBidiChar*)(stringSrc.c_str() + lineStart), lineLen, &baseCopy, visual, NULL, NULL, NULL))
+    if (fribidi_log2vis((const FriBidiChar*)(stringSrc.c_str() + lineStart), lineLen, &baseCopy,
+      visual, logicalToVisualMap, NULL, NULL))
```
Align to other arguments, space on cast and nullptr.

```cpp
if (fribidi_log2vis((const FriBidiChar*) (stringSrc.c_str() + lineStart),
                    lineLen, &baseCopy, visual, logicalToVisualMap, nullptr, nullptr))
```

I have no idea what the correct c++ cast is there. Maybe `reinterpret_cast`? Try it and see.

phunkyfish commented on xbmc/utils/CharsetConverter.cpp:
```diff
@@ -676,9 +681,12 @@ bool CCharsetConverter::utf32ToW(const std::u32string& utf32StringSrc, std::wstr
 }
 
-bool CCharsetConverter::utf32logicalToVisualBiDi(const std::u32string& logicalStringSrc, std::u32string& visualStringDst, bool forceLTRReadingOrder /*= false*/, bool failOnBadString /*= false*/)
+bool CCharsetConverter::utf32logicalToVisualBiDi(const std::u32string& logicalStringSrc,
+  std::u32string& visualStringDst, bool forceLTRReadingOrder /*= false*/,
+  bool failOnBadString /*= false*/, int* logicalToVisualMap /*= nullptr*/)
```
Alignment:

```cpp
bool CCharsetConverter::utf32logicalToVisualBiDi(const std::u32string& logicalStringSrc,
                                                 std::u32string& visualStringDst,
                                                 bool forceLTRReadingOrder /*= false*/,
                                                 bool failOnBadString /*= false*/,
                                                 int* logicalToVisualMap /*= nullptr*/)
{
  return CInnerConverter::logicalToVisualBiDi(
      logicalStringSrc, visualStringDst, 
      forceLTRReadingOrder ? FRIBIDI_TYPE_LTR : FRIBIDI_TYPE_PDF,
      failOnBadString, logicalToVisualMap);
}
```

phunkyfish commented on xbmc/utils/CharsetConverter.h:
```diff
@@ -98,9 +98,12 @@ class CCharsetConverter : public ISettingCallback
    * @return true on success, false otherwise
    */
-  static bool utf32logicalToVisualBiDi(const std::u32string& logicalStringSrc, std::u32string& visualStringDst, bool forceLTRReadingOrder = false, bool failOnBadString = false);
+  static bool utf32logicalToVisualBiDi(const std::u32string& logicalStringSrc,
+    std::u32string& visualStringDst, bool forceLTRReadingOrder = false,
+    bool failOnBadString = false, int* logicalToVisualMap = nullptr);
```
Alignment:

```cpp
static bool utf32logicalToVisualBiDi(const std::u32string& logicalStringSrc,
                                     std::u32string& visualStringDst,
                                     bool forceLTRReadingOrder = false,
                                     bool failOnBadString = false,
                                     int* logicalToVisualMap = nullptr);
```

phunkyfish reviewed (changes requested):
Nearly there ;)

Mainly just alignment issues outside of the cast and `!` conditional. Once you add you code in git you can just run `git-clang-format`, add those changes again and then commit if you find that easier than following my review.

The only thing I'm not sure of is how to do the C style cast as a C++ one.

AhmedAdelAbouzeid commented:
I used clang for formatting and replaced the C style cast with reinterpret_cast, please recheck. Thanks.

phunkyfish commented:
It's looking good, can you squash it all down to a single commit matching the PR title?

AhmedAdelAbouzeid commented:
All changes are part of a single commit now. Please re-check, Thanks!

phunkyfish reviewed (approved):
Nice, the code is ready. It went into a test build yesterday, let's give a few days and we can ask @MilhouseVH  if there has been any feedback.

Well done!

AhmedAdelAbouzeid commented on xbmc/guilib/GUITextLayout.cpp:
```diff
@@ -256,56 +256,68 @@ void CGUITextLayout::UpdateStyled(const vecText &text, const std::vector<UTILS::
-        flippedText.push_back(sectionStyle | sectionFlipped[j]);
+      for (unsigned int j = 0; j < visualText.size(); j++)
+      {
+        styledVisualText.push_back(style[logicalToVisualMap[j]] | visualText[j]);
```
I just realized that there is a mistake here. I used the value of logicalToVisualMap[j] as index of style[] which is in logical order.  While this works for most cases, it's not correct; it should be visualToLogicalMap[] instead.
```suggestion
        styledVisualText.push_back(style[visualToLogicalMap[j]] | visualText[j]);
```
To correct this I will rename logicalToVisualMap to visualToLogicalMap everywhere it's used and use the respective output from fribidi for it.

AhmedAdelAbouzeid committed:
Prevent RTL text from mirrorring around cursor for on-screen keyboard

1- Revert the character used for cursor blinking for on-screen keyboard to the original.

2- BidiTransform() function was re-written such that calls to FriBidi are performed for a complete line at a time rather than breaking it for differently styled sections (RTL text mirroring around styled characters occurred due to this line splitting). The styling is reapplied to the line after FriBidi call.

3- Visual to logical string mapping was added as an optional parameter for functions wrapping FriBidi calls to support functionality needed in point 2.

AhmedAdelAbouzeid commented:
I corrected this (changed logicalToVisualMap to visualToLogicalMap), no other changes were made.

phunkyfish commented:
Great, thanks @AhmedAdelAbouzeid

phunkyfish commented:
Will merge this tomorrow.

yeahme49 commented:
Ran into a possible bug with this commit. Was watching a video using the YouTube addon and after the video finished and it should have brought me back to the search listing page, Kodi crashed. Nothing about the crash in kodi.log but opening the .dmp file with Visual Studio this is what shows up
![crash](https://user-images.githubusercontent.com/58564160/82736843-182f6780-9cf2-11ea-821e-7ed8bb957cb7.PNG)

Any ideas or any other info needed?

AhmedAdelAbouzeid commented:
@yeahme49 Is this reproducible ? if so then please share the search string and title of the video that causes the crash. Also, can you please upload the dmp file ?

yeahme49 commented:
I was able to reproduce it. If you search 'LTT boiling' there should be a video titled 'CPU Cooling with BOILING LIQUID'. The first time this video shows up in a search, it works fine. But any subsequent times it is in the list, Kodi crashes. The actual title of the video is 'CPU Cooling with BOILING LIQUID 🔥🔥' except the emoji flames don't show up in Kodi. I'm guessing it's the emojis that are causing the issue.
[kodi_crashlog-19.0-ALPHA1 Git_20200521-3f2e051032-20200523-142758.dmp.zip](https://github.com/xbmc/xbmc/files/4672417/kodi_crashlog-19.0-ALPHA1.Git_20200521-3f2e051032-20200523-142758.dmp.zip)

AhmedAdelAbouzeid commented:
@yeahme49 Thanks for pointing this out. The emojis indeed triggered the issue leading to the crash.
Fixed in  #17932
